### PR TITLE
Wait for amp selector to load before submissions

### LINKED ISSUE #6927: amp-form should wait for amp-selector to finish loading before submission
This is a followup issue to keep track of `amp-selector` and `amp-form` submission.

Splitting these into different PRs. First PR will address waiting for `amp-form` through delegating XHR submissions through action service and the next PR will address waiting for `amp-selector`.

### PR DESCRIPTION
**Pull request by mkhatib**
Fixes #6927 

This depends on #6929

### CODE REVIEW

File: extensions/amp-form/0.1/amp-form.js [hunk 1]
```diff
@@ -147,12 +160,36 @@ export class AmpForm {
...
+   * Returns a promise that will be resolved when all dependencies used inside the form
+   * tag are loaded (e.g. amp-selector) or 500ms timeout - whichever is first.
+   * @return {!Promise}
+   * @private
+   */
+  whenDependenciesReady_() {
+    const usedDeps = this.form_.querySelectorAll(EXTERNAL_DEPS.join(','));
+    const depsDict = {};
+    const depsPromises = [];
+    for (let i = usedDeps.length - 1; i >= 0; i--) {
```
Comment (camelburrito) [inline extensions/amp-form/0.1/amp-form.js]: you should not have to loop through all matching elements. 

You can loop through EXTERNAL_DEPS and check of getElementsByTagName(tag).length > 0 then wait for the tag.

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: I guess now we only have one dependency `getElementsByTagName` is fine. Not sure if `querySelectorAll` would be faster than multiple `getElementsByTagName` in the future.

Will update this to use `getElementsByTagName`.

Comment (camelburrito) [inline extensions/amp-form/0.1/amp-form.js]: you will still have to do a tagName read with queryseletorall in a loop (atleast one per the tag in the array). I still think what i suggested is slightly better and clean

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

**Comment by mkhatib** [comment]
@camelburrito PTAL 👀

File: extensions/amp-form/0.1/amp-form.js [hunk 2]
```diff
@@ -155,14 +168,40 @@ export class AmpForm {
...
+   * @private
+   */
+  whenDependenciesReady_() {
+    const depsDict = {};
+    const depsPromises = [];
+    for (let i = EXTERNAL_DEPS.length - 1; i >= 0; i--) {
+      const tagName = EXTERNAL_DEPS[i];
+      const depIsUsed = this.form_.getElementsByTagName(tagName).length !== 0;
+      if (depIsUsed && depsDict[tagName] === undefined) {
+        // TODO(mkhatib, #7032): Implement a way to wait for an element to be built to
```
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: @dvoytenko I prefer to do this in a separate PR if that's ok. lmk.

File: extensions/amp-form/0.1/amp-form.js [hunk 3]
```diff
@@ -155,14 +168,40 @@ export class AmpForm {
     /** @const @private {!./form-validators.FormValidator} */
     this.validator_ = getFormValidator(this.form_);
 
-    // TODO(mkhatib, #6927): Wait for amp-selector to finish loading if the current form
-    // is using it.
-    this.actions_.installActionHandler(
-        this.form_, this.actionHandler_.bind(this));
+    // Wait for amp-selector to finish loading if the current form is using it.
+    this.whenDependenciesReady_().then(() => {
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: I'm not sure it makes sense to wait to install action handler here, esp since we have other listeners that could override the behavior. Should, instead, action handler be installed right away, but block on dependencies?

File: extensions/amp-form/0.1/amp-form.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: I think that the implementation would be drastically different and we should consider doing build promise work first.

File: extensions/amp-form/0.1/amp-form.js [hunk 4]
```diff
@@ -563,6 +602,10 @@ function updateInvalidTypesClasses(element) {
  * @return {boolean} Whether the element is valid or not.
  */
 function checkUserValidity(element, propagate = false) {
+  // If this is not a field type with checkValidity don't do anything.
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Is the assumption that `amp-selector` somehow will implement this interface? Do we have issues that would track this work? In particular `required` is very sensible for `amp-selector`.

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Yes I filed an issue to track this #6930

File: extensions/amp-form/0.1/amp-form.js (hunk 3 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: > esp since we have other listeners that could override the behavior

You mean other `submit` listeners? This would still be very similar situation with blocking on dependencies there are still other listeners.

If you meant other action handlers yeah if some other code install an action listener on the form element these `submit` events will be gone from the queue.

File: extensions/amp-form/0.1/amp-form.js (hunk 2 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: sgtm. will leave this PR until I have something for the build promise.
